### PR TITLE
[Dynamic buffer calc] Support dynamic buffer calculation

### PR DESCRIPTION
**Pull request by stephenxs**
**- Why I did it**
To support dynamic buffer calculation.
This PR also depends on the following PRs for sub modules
- [sonic-swss: [buffermgr/bufferorch] Support dynamic buffer calculation #1338](https://github.com/Azure/sonic-swss/pull/1338)
- [sonic-swss-common: Dynamic buffer calculation #361](https://github.com/Azure/sonic-swss-common/pull/361)
- [sonic-utilities: Support dynamic buffer calculation #973](https://github.com/Azure/sonic-utilities/pull/973)

**- How I did it**
1. add tables required for the feature:
 - ASIC_TABLE in files/build_templates/asic_table.j2
 - PERIPHERAL_TABLE in files/build_templates/peripheral_table.j2
 - PORT_PERIPHERAL_TABLE on a per-platform basis in device/mellanox/x86_64-mlnx_msn3800-r0/port_peripheral_config.j2
 - DEFAULT_LOSSLESS_BUFFER_PARAMETER and LOSSLESS_TRAFFIC_PATTERN in files/build_templates/buffers_config.j2
 - Add lossless PGs (3-4) for each port in files/build_templates/buffers_config.j2
2. copy the newly introduced j2 files into image and rendering them when system starts
3. update the CLI options for buffermgrd so that it can start with dynamic mode
4. Adjust the order in which swss daemons start, making buffermgrd start before orchagent.
   This is to make sure buffermgrd can get everything ready before orchagent starts especially during warm reboot.
5. Optimize the way in which orchagent fetches the asic vendor name:
   - fetch the vendor name when creates the docker and pass it as a docker environment variable
   - orchagent and buffermgrd can use this passed-in variable
6. Clear buffer related tables from STATE_DB when swss docker starts
7. Update the src/sonic-config-engine/tests/sample_output/buffers-dell6100.json according to the buffer_config.j2
8. Remove buffer pool sizes for ingress pools and egress_lossy_pool
   Update the buffer settings for dynamic buffer calculation

Signed-off-by: Stephen Sun <stephens@nvidia.com>

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: src/sonic-config-engine/tests/sample_output/buffers-dell6100.json [hunk 1]
```diff
@@ -108,134 +108,266 @@
         "Ethernet0|0": {
             "profile" : "[BUFFER_PROFILE|ingress_lossy_profile]"
         },
+        "Ethernet0|3-4": {
+            "NULL" : "NULL"
```
Comment (neethajohn) [inline src/sonic-config-engine/tests/sample_output/buffers-dell6100.json:112]: need to change all occurrences below to "profile": "NULL"

**Comment by neethajohn** [comment]
Can you also include testcases to ensure that the asic_table.j2 and peripheral_table.j2 are rendered properly?

File: dockers/docker-orchagent/supervisord.conf [hunk 2]
```diff
@@ -26,6 +26,16 @@ stdout_logfile=syslog
...
 dependent_startup=true
 
+[program:buffermgrd]
+command=/usr/bin/buffermgrd -l /usr/share/sonic/hwsku/pg_profile_lookup.ini -a /etc/sonic/asic_table.json -p /etc/sonic/peripheral_table.json
+priority=2
+autostart=false
+autorestart=false
+stdout_logfile=syslog
+stderr_logfile=syslog
+dependent_startup=true
```
[the remark this answers is not in the document]
Comment (stephenxs) [reply dockers/docker-orchagent/supervisord.conf]: Yes, we can.
Done.

File: src/sonic-config-engine/tests/sample_output/buffers-dell6100.json (hunk 1 above)
Comment (stephenxs) [reply src/sonic-config-engine/tests/sample_output/buffers-dell6100.json]: Fixed

Comment (stephenxs) [reply src/sonic-config-engine/tests/sample_output/buffers-dell6100.json]: Fixed

File: files/build_templates/asic_table.j2 [hunk 3]
```diff
@@ -0,0 +1,69 @@
+{%- if DEVICE_METADATA is defined and DEVICE_METADATA['localhost']['platform'] is defined %}
+{%-     set platform = DEVICE_METADATA['localhost']['platform'] %}
+{%- else -%}
+{%-     set platform = "vs-platform" %}
+{%- endif -%}
+
+
+[
+{% set platform2asic = {
```
[the remark this answers is not in the document]
Comment (stephenxs) [reply files/build_templates/asic_table.j2:9]: This is the way we agreed on during the design review.
But if this is a concern, I'll do so, moving the files to the platform subdir.

File: device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/buffers_defaults_t0.j2 [hunk 4]
```diff
@@ -1,8 +1,5 @@
 {% set default_cable = '5m' %}
-{% set ingress_lossless_pool_size =  '4580864' %}
-{% set ingress_lossy_pool_size =  '4580864' %}
 {% set egress_lossless_pool_size =  '13945824' %}
-{% set egress_lossy_pool_size =  '4580864' %}
```
[the remark this answers is not in the document]
Comment (stephenxs) [reply device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/buffers_defaults_t0.j2]: Fixed by using the new upgrading flow.

File: device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/buffers_defaults_t1.j2 [hunk 5]
```diff
@@ -1,8 +1,5 @@
 {% set default_cable = '5m' %}
-{% set ingress_lossless_pool_size =  '3302912' %}
-{% set ingress_lossy_pool_size =  '3302912' %}
 {% set egress_lossless_pool_size =  '13945824' %}
-{% set egress_lossy_pool_size =  '3302912' %}
```
[the remark this answers is not in the document]
Comment (stephenxs) [reply device/mellanox/x86_64-mlnx_msn2700-r0/ACS-MSN2700/buffers_defaults_t1.j2]: Fixed by using the new upgrading flow.

File: files/build_templates/asic_table.j2 (hunk 3 above)
[the remark this answers is not in the document]
Comment (stephenxs) [reply files/build_templates/asic_table.j2]: fixed.

File: dockers/docker-orchagent/supervisord.conf [hunk 6]
```diff
@@ -124,14 +124,14 @@ dependent_startup=true
...
-command=/usr/bin/buffermgrd -l /usr/share/sonic/hwsku/pg_profile_lookup.ini
+command=/usr/bin/buffermgrd.sh
 priority=11
 autostart=false
 autorestart=false
 stdout_logfile=syslog
 stderr_logfile=syslog
 dependent_startup=true
-dependent_startup_wait_for=swssconfig:exited
+dependent_startup_wait_for=rsyslogd:running
```
[the remark this answers is not in the document]
Comment (stephenxs) [reply dockers/docker-orchagent/supervisord.conf:134]: Possibly the leftover of an old change which was already reverted.
Will fix it

[the remark this answers is not in the document]
Comment (stephenxs) [reply dockers/docker-orchagent/supervisord.conf]: Fixed.
Also rebased the PR based on a recent sonic-buildimage.

**Review by neethajohn — APPROVED** [review]

File: platform/vs/docker-sonic-vs/Dockerfile.j2 [hunk 7]
```diff
@@ -157,6 +157,8 @@ COPY ["port_breakout_config_db.json", "/etc/sonic/"]
 COPY ["database_config.json", "/etc/default/sonic-db/"]
 COPY ["hostname.j2", "/usr/share/sonic/templates/"]
 COPY ["default_chassis_cfg.json", "/etc/default/sonic-db/"]
+COPY ["asic_table.json", "/etc/sonic/"]
```
Comment (neethajohn) [inline platform/vs/docker-sonic-vs/Dockerfile.j2:160]: What about the peripheral_table for vs?

Comment (stephenxs) [reply platform/vs/docker-sonic-vs/Dockerfile.j2:160]: If there is no gear box installed we don’t need that

Comment (neethajohn) [reply platform/vs/docker-sonic-vs/Dockerfile.j2:160]: In the docker-sonic-vs/buffermgrd.sh, 'peripheral_table.json' is one of the mandatory args for starting buffermgr in dynamic mode

Comment (stephenxs) [reply platform/vs/docker-sonic-vs/Dockerfile.j2:160]: Will fix it.

**Review by neethajohn — APPROVED** [review]

**Comment by stephenxs** [comment]
retest this, please

**Comment by qiluo-msft** [comment]
This PR is wrong merged and we manually remove the commit from master branch. Discussing with PR owner to submit another one.

**Pull request merged**